### PR TITLE
ENH: Validate AsOfJoin tolerance and attempt interval unit conversion

### PR DESCRIPTION
**Pull request by emilyreff7**
Unlike other operations, not all of the arguments of AsOfJoin are automatically passed through the validation chain that occurs in superclass Annotable, since they are not passed into super(). The result is that whereas other operations with offset-like arguments (e.g. Lead/Lag) have their offsets' dtypes automatically inferred, the same does not occur for 'tolerance' in AsOfJoin. We can simply wire this validation in ourselves.

Additionally, the default unit for Interval is 's' regardless of the unit of the value of origin. We can make a better attempt at inferring unit before immediately defaulting to seconds.

### CODE REVIEW

File: ibis/expr/tests/test_table.py [hunk 1]
```diff
@@ -695,6 +698,89 @@ def test_asof_join_with_by():
...
+            pd.Timedelta('3', unit='Y'),
+            id='years',
+            marks=pytest.mark.xfail(
+                reason='Year conversion from Timedelta to ibis interval '
+                'not supported'
+            ),
+        ),
+    ],
+)
+def test_asof_join_with_tolerance(ibis_interval, timedelta_interval):
```
Comment (toryhaavik) [inline ibis/expr/tests/test_table.py]: I wonder if we can break these tests out into separate parts. testing the interval translation (probably moved to test_datatypes.py) and testing the asof_join with tolerance (here).

File: ibis/expr/datatypes.py [hunk 2]
```diff
@@ -32,6 +32,7 @@
 try:
     if sys.version_info >= (3, 6):
         import shapely.geometry
+
```
Comment (toryhaavik) [inline ibis/expr/datatypes.py]: @xmnlab @cpcloud gotta figure out how to better enforce that black is run on every commit, so we don't have these extraneous formatting changes in more recent reviews.

Comment (xmnlab) [reply ibis/expr/datatypes.py]: I think now black is working properly .. but as it was not working properly in the last months ... some files will be changed now for black. I will open a PR just for black formatting.

**Comment by toryhaavik** [comment]
hopefully #1963 can get in ahead of this, you can rebase on top of it, and then only your enhancement will be there, not the miscellaneous formatting changes too

**Comment by toryhaavik** [comment]
@emilyreff7 please rebase on top of master and then i'll approve!

**Comment by emilyreff7** [comment]
> @emilyreff7 please rebase on top of master and then i'll approve!

Just rebased, looks like only my changes remain :)

**Comment by toryhaavik** [comment]
these seem like spurious failures, maybe try amending your commit and force pushing to kick off CI again

**Comment by emilyreff7** [comment]
> these seem like spurious failures, maybe try amending your commit and force pushing to kick off CI again

looks good now!

**Comment by emilyreff7** [comment]
> > these seem like spurious failures, maybe try amending your commit and force pushing to kick off CI again
> 
> looks good now!

nvm, accidentally closed so the tests are rerunning now :(

**Comment by emilyreff7** [comment]
> > > these seem like spurious failures, maybe try amending your commit and force pushing to kick off CI again
> > 
> > 
> > looks good now!
> 
> nvm, accidentally closed so the tests are rerunning now :(

okay, now all good

**Pull request merged**